### PR TITLE
feat: prune layouts based on stats

### PR DESCRIPTION
**Pull request by danking**
LayoutReaders can now check if a given row range is fully pruned by their Scan expression (which must be Boolean-valued). Moreover, PrunedSplitIterator uses this information to avoid even generating RowMasks for ranges which do not contain any matching rows.

### CODE REVIEW

**Comment by danking** [comment]
On the most selective queries, we see a dramatic improvement:
```
query              develop                          this PR
clickbench/q-37    104.59 885.7±385.63ms            1.00      8.5±0.09ms        
clickbench/q-38    120.35 976.5±483.77ms            1.00      8.1±0.10ms        
clickbench/q-39    100.28 782.6±448.30ms            1.00      7.8±0.04ms        
clickbench/q-40    103.26 924.9±457.85ms            1.00      9.0±0.16ms        
clickbench/q-41    1.01  697.3±331.13ms             1.00  688.8±306.88ms        
clickbench/q-42    1.60  880.2±444.12ms             1.00  549.1±218.58ms        
clickbench/q-43    1.13  842.3±391.87ms             1.00  745.3±415.63ms        
```

**Comment by danking** [comment]
Somewhat selective queries do OK:
```
query              develop                          this PR
clickbench/q-22    1.52       5.3±0.09s             1.00       3.5±0.26s        
clickbench/q-23    5.52      10.8±0.23s             1.00  1962.3±155.42ms        
clickbench/q-24    1.00      39.3±0.40s             1.00      39.3±0.57s        
clickbench/q-25    1.21       2.4±0.21s             1.00  1946.6±77.83ms        
```

**Comment by danking** [comment]
random access is now nearly as fast as it was before https://github.com/spiraldb/vortex/commit/520ffc4569f3f9070eeea7e1031d8ccd72adf668

File: vortex-file/src/read/layouts/columnar.rs [hunk 1]
```diff
@@ -343,6 +346,39 @@ impl LayoutReader for ColumnarLayoutReader {
...
+            .filter(|(_, a)| a.is_none())
+        {
+            match self.children[i].can_prune(begin, end)? {
+                PruningRead::ReadMore(message) => messages.extend(message),
+                PruningRead::CanPrune(is_pruned) => *can_prune_child = Some(is_pruned),
+            }
+        }
+
+        if messages.is_empty() {
+            let any_child_is_pruned = in_progress_guard
```
Comment (lwwmanning) [inline vortex-file/src/read/layouts/columnar.rs:379]: big comment here plz

**Review by lwwmanning — APPROVED** [review]

**Pull request merged**